File: queuebystack.java
import java.util.Stack;

class Queue<T> {
	private Stack<T> s;

	// Constructor
	Queue() {
		s = new Stack<>();
	}

	// Add an item to the queue
	public void enqueue(T data)
	{
		// push item into the first stack
		s.push(data);
	}

	// Remove an item from the queue
	public T dequeue()
	{
		// if the stack is empty
		if (s.isEmpty())
		{
			System.out.println("Underflow!!");
			System.exit(0);
		}

		// pop an item from the stack
		T top = s.pop();

		// if the stack becomes empty, return the popped item
		if (s.isEmpty()) {
			return top;
		}

		// recur
		T item = dequeue();

		// push popped item back into the stack
		s.push(top);

		// return the result of dequeue() call
		return item;
	}
}

class Main
{
	public static void main(String[] args)
	{
		int[] keys = { 5, 4, 3, 2, 1 };
		Queue<Integer> q = new Queue<Integer>();

		// insert the above keys into the queue
		for (int key: keys) {
			q.enqueue(key);
		}

		System.out.println(q.dequeue());
		System.out.println(q.dequeue());	
	}
}

File: queuebystack.cpp
#include <bits/stdc++.h> 
using namespace std; 

struct Queue { 
	stack<int> stack1, stack2; 

	void enqueue(int x) 
	{ 
		// Move all elements from s1 to s2 
		while (!stack1.empty()) { 
			stack2.push(stack1.top()); 
			stack1.pop(); 
		} 

		// Push item into s1 
		stack1.push(x); 

		// Push everything back to s1 
		while (!stack2.empty()) { 
			stack1.push(stack2.top()); 
			stack2.pop(); 
		} 
	} 

	// Dequeue an item from the queue 
	int dequeue() 
	{ 
		// if first stack is empty 
		if (stack1.empty()) { 
			cout << "queue is Empty"; 
			exit(0); 
		} 

		// Return the top of stack1 
		int x = stack1.top(); 
		stack1.pop(); 
		return x; 
	} 
}; 

int main() 
{ 
	Queue q; 
	q.enqueue(3); 
	q.enqueue(4); 
	q.enqueue(5); 

	cout << q.dequeue() << endl; 
	cout << q.dequeue() << endl; 
	cout << q.dequeue() << endl; 

	return 0; 
} 
